File: scala/86.scala
object Solution {
    def partition(head: ListNode, x: Int): ListNode = {
        val less = ListNode()
        val greater = ListNode()
        var lessTail = less
        var greaterTail = greater
        var ptr = head
        while(ptr != null) {
          if (ptr.x < x) {
            lessTail.next = ptr
            ptr = ptr.next
            lessTail.next.next = null
            lessTail = lessTail.next
          } else {
            greaterTail.next = ptr
            ptr = ptr.next
            greaterTail.next.next = null
            greaterTail = greaterTail.next
          }
        }
        lessTail.next = greater.next
        less.next
    }
}

File: leetcodeJava/86.java
package leetcodeJava;

class Solution {
    public ListNode partition(ListNode head, int x) {
        if (head == null) return null;
        ListNode l1 = new ListNode(), l2 = new ListNode();
        ListNode t1 = l1, t2 = l2;
        var ptr = head;
        while(ptr != null) {
            if (ptr.val < x) {
                t1.next = new ListNode(ptr.val);
                t1 = t1.next;
            } else {
                t2.next = new ListNode(ptr.val);
                t2 = t2.next;
            }
            ptr = ptr.next;
        }
        t1.next = l2.next;
        return l1.next;
    }
}
